### PR TITLE
[SPARK-27350][SQL] Support create table on data source V2

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?

This patch adds support:

- create table on data source v2

- insert into data source v2 relation

- insert into data source v2 directory.

## How was this patch tested?

new UT tests.

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Hi @cloud-fan, would you mind take a look? Thanks a lot.

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
ok to test

**Review by uncleGen — COMMENTED** [review]
cc @cloud-fan Left a few comments.

**Comment by uncleGen** [comment]
@ConeyLiu are you still working on this? If not may I ask to close? Others can continue to work on this.

**Comment by ConeyLiu** [comment]
Hi @uncleGen, I'm sorry for the later reply. Will address the comments recently.

**Comment by dilipbiswal** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala [hunk 1]
```diff
@@ -53,25 +50,6 @@ case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with Ca
       CreateTable(tableDesc, mode, Some(query))
   }
 
-  object V1WriteProvider {
-    private val v1WriteOverrideSet =
```
[the remark this answers is not in the document]
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala]: The `V1WriteProvider` will remove the provider message when it is v2 source. And the checking is done in the command.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FindDataSourceV2Table.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        Option(tableMeta.schema),
+        tableMeta.provider.get,
+        pathOption.toMap)
+
+      if (shouldReadInV2) {
+        readDataSourceTable(tableMeta)
+      } else {
+        unresolved
+      }
+  }
```
[the remark this answers is not in the document]
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FindDataSourceV2Table.scala]: Hive table will be handled in the `FindDataSourceTable`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala [hunk 3]
```diff
@@ -35,10 +42,15 @@ private[sql] object DataSourceV2Utils extends Logging {
...
+   * @return an case insensitive immutable map that contains all the extracted and transformed
+   *         k/v pairs.
    */
-  def extractSessionConfigs(source: TableProvider, conf: SQLConf): Map[String, String] = {
-    source match {
+  def extractSessionConfigs(
+      source: TableProvider,
+      conf: SQLConf,
+      extraOptions: Map[String, String]): CaseInsensitiveStringMap = {
+    val extracted = source match {
```
[the remark this answers is not in the document]
Comment (ConeyLiu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Utils.scala]: After calling `extractSessionConfigs`, we still need to merge the return `extraOptions` and transfer it to `CaseInsensitiveStringMap`. This change is just a simplify and I have updated the doc.

**Comment by HyukjinKwon** [comment]
I think it's being taken over by https://github.com/apache/spark/pull/24617. Shall we close if I am correct?

**Comment by ConeyLiu** [comment]
close it.

**Pull request closed without merging**